src/relcast.erl

```diff
@@ -1050,11 +1069,111 @@ make_seq(ID, #state{seq_map=SeqMap, ids=A}=State) ->
 reset_seq(ID, #state{seq_map=SeqMap}=State) ->
     State#state{seq_map=maps:remove(ID, SeqMap)}.
 
-maybe_serialize(_Mod, Old, New, _Transaction) when Old == New ->
-    false;
-maybe_serialize(Mod, _Old, New, Transaction) ->
-    ok = rocksdb:transaction_put(Transaction, ?stored_module_state, Mod:serialize(New)),
-    true.
+maybe_serialize(_Mod, Old, #state{modulestate = New} = S) when Old == New ->
+    S;
+maybe_serialize(Mod, _Old, #state{modulestate = New0,
+                                  old_serialized = Old,
+                                  transaction = Transaction} = S) ->
+    New = Mod:serialize(New0),
+    KeyTree = do_serialize(Mod, Old, New, ?stored_key_prefix, Transaction),
+    S1 = maybe_write_key_tree(KeyTree, S),
+    S1#state{old_serialized = New, transaction_dirty = true}.
+
+maybe_write_key_tree(KeyTree, S) when KeyTree == S#state.key_tree->
+    S;
+maybe_write_key_tree(KeyTree, S) ->
+    ok = rocksdb:transaction_put(S#state.transaction, ?stored_key_tree,
+                                 term_to_binary(KeyTree, [compressed])),
+    S#state{key_tree = KeyTree}.
+
+old_size(M) when is_map(M) ->
+    maps:size(M);
+old_size(_) ->
+    -1.
+
+do_serialize(Mod, Old, New, Prefix, Transaction) ->
+    case New of
+        State when is_binary(State) ->
+            ok = rocksdb:transaction_put(Transaction, ?stored_module_state, State),
+            bin;
+        StateMap ->
+            S = lists:sort(maps:to_list(StateMap)),
+            SSize = maps:size(StateMap),
+            OSize = old_size(Old),
+            O = case Old of
+                    %% since we're serializing for the first time, we need to make sure that
+                    %% everything gets written out, otherwise we have partial state that
+                    %% won't restore correctly.
+                    BorU when BorU == undefined orelse
+                              is_binary(BorU) ->
+                        lists:map(fun({K, _V}) -> {K, never_ever_match_with_anything} end, S);
+                    _Size when OSize =/= SSize ->
+                        SKeys = maps:keys(StateMap),
+                        OKeys = maps:keys(Old),
+                        Old1 = lists:foldl(fun(L, OM) ->
+                                                   OM#{L => undefined}
+                                           end,
+                                           Old,
+                                           SKeys -- OKeys),
+                        maps:to_list(maps:without(OKeys -- SKeys, Old1));
+                    _ ->
+                        lists:sort(maps:to_list(Old))
+                end,
+            L = lists:zip(S, O),
+            KeyTree =
+                lists:map(
+                  fun({{K, V}, {_, V}}) ->
```

Comment:
This clause is not covered by the test code, is this when the old and new value is the same?

Reply:
yes.